### PR TITLE
[MIN-60] Update load data from DVC step into generic step

### PR DESCRIPTION
**Pull request by Christopher-Norman**
This PR converts the `load_data` step in the data preparation pipeline into a generic step to be used within the data preparation pipeline and the data embedding pipeline.

This step pulls the specified tag/hash `data` folder using Git and then pulls with DVC to obtain the specified data version.

### CODE REVIEW

File: utils/data_version_control.py [hunk 1]
```diff
@@ -98,3 +99,14 @@ def version_new_data(filename_roots: List[str]) -> None:
...
+
+def git_checkout_folder(tag_name: str, folder_name: str) -> None:
+    """Checkout a specified folder within a tagged commit or commit hash on Git.
+
+    Args:
+        tag_name (str): Git tag or commit hash to checkout.
+        folder_name (str): Folder name to checkout e.g. 'data'.
+    """
+    g = git.cmd.Git("")
+    g.checkout(tag_name, folder_name)
```
Comment (osw282) [inline utils/data_version_control.py]: Do we assume that the tag will always exists? What happens if it doesn't?

**Review by osw282 — COMMENTED** [review]
Looks good overall, just one question. Nice work. 🚀

File: utils/data_version_control.py (hunk 1 above)
Comment (Christopher-Norman) [reply utils/data_version_control.py]: Adding checks now

**Review by swells2020 — APPROVED** [review]
LGTM 🐮

File: pipelines/data_preparation_pipeline/config_data_preparation_pipeline.yaml [hunk 2]
```diff
@@ -1,6 +1,9 @@
 steps:
   load_data:
     enable_cache: False
+    parameters:
```
Comment (JonoCX) [inline pipelines/data_preparation_pipeline/config_data_preparation_pipeline.yaml:4]: General question: is this consistent between pipelines?

File: steps/generic_steps/load_data_step/load_data_step.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                Name: timestamp, dtype: datetime64[ns]
+                Name: url, dtype: object
+    """
+    git_checkout_folder(tag_name=data_version, folder_name="data")
+    pull_data()
+
+    mind_file_path = os.path.join(DATA_DIR, f"mind_data_{data_postfix}.csv")
+    nhs_file_path = os.path.join(DATA_DIR, f"nhs_data_{data_postfix}.csv")
+
+    files_exist(
```
Comment (JonoCX) [inline steps/generic_steps/load_data_step/load_data_step.py:46]: There should be a test for this execution path

File: utils/data_version_control.py [hunk 4]
```diff
@@ -98,3 +102,63 @@ def version_new_data(filename_roots: List[str]) -> None:
...
+        g = git.cmd.Git(PROJECT_ROOT_DIR)
+        g.checkout(tag_name, folder_name)
+    elif commit_hash is not None and git_commit_hash_exists(commit_hash):
+        g = git.cmd.Git(PROJECT_ROOT_DIR)
+        g.checkout(commit_hash, folder_name)
+    else:
+        logger.error("Error, you must specify a valid tag or commit hash.")
+
+
+def git_tag_exists(tag_name: str) -> bool:
```
Comment (JonoCX) [inline utils/data_version_control.py]: These should go before `git_checkout_folder` to follow the order of execution and should marked as private functions

File: utils/data_version_control.py [hunk 5]
```diff
@@ -98,3 +102,63 @@ def version_new_data(filename_roots: List[str]) -> None:
     dvc_files = [f"{fname}.csv.dvc" for fname in filename_roots]
     add_csv_files_to_dvc(csv_files)
     add_and_commit_dvc_files_to_git(dvc_files)
+
+
+def git_checkout_folder(
+    tag_name: Optional[str] = None,
+    commit_hash: Optional[str] = None,
+    folder_name: str = "data",
+) -> None:
```
Comment (JonoCX) [inline utils/data_version_control.py]: Should this return something, perhaps a `Path` object which is a pointer to the checked out folder?

File: utils/data_version_control.py [hunk 6]
```diff
@@ -98,3 +102,63 @@ def version_new_data(filename_roots: List[str]) -> None:
...
+        folder_name (str): Folder name to checkout e.g. 'data'.
+    """
+    if tag_name is not None and git_tag_exists(tag_name):
+        g = git.cmd.Git(PROJECT_ROOT_DIR)
+        g.checkout(tag_name, folder_name)
+    elif commit_hash is not None and git_commit_hash_exists(commit_hash):
+        g = git.cmd.Git(PROJECT_ROOT_DIR)
+        g.checkout(commit_hash, folder_name)
+    else:
+        logger.error("Error, you must specify a valid tag or commit hash.")
```
Comment (JonoCX) [inline utils/data_version_control.py]: Should we want to halt execution at this point?

File: utils/data_version_control.py [hunk 7]
```diff
@@ -98,3 +102,63 @@ def version_new_data(filename_roots: List[str]) -> None:
...
+def git_tag_exists(tag_name: str) -> bool:
+    """Checks for the existance of a given Git tag name.
+
+    Args:
+        tag_name (str): Name of tag to check existence.
+
+    Returns:
+        bool: True, if Git tag exists. False otherwise.
+    """
+    try:
```
Comment (JonoCX) [inline utils/data_version_control.py:116]: This (and the function below) could be simplified: `return tag_name in git.Repo().tags.keys()`

File: utils/data_version_control.py [hunk 8]
```diff
@@ -98,3 +102,63 @@ def version_new_data(filename_roots: List[str]) -> None:
     dvc_files = [f"{fname}.csv.dvc" for fname in filename_roots]
     add_csv_files_to_dvc(csv_files)
     add_and_commit_dvc_files_to_git(dvc_files)
+
+
+def git_checkout_folder(
```
Comment (JonoCX) [inline utils/data_version_control.py]: There needs to be tests added (happy and unhappy path) for these three functions

**Review by JonoCX — CHANGES_REQUESTED** [review]
Nice - see comments 🦕

File: pipelines/data_preparation_pipeline/config_data_preparation_pipeline.yaml (hunk 2 above)
Comment (Christopher-Norman) [reply pipelines/data_preparation_pipeline/config_data_preparation_pipeline.yaml:4]: No

File: utils/data_version_control.py (hunk 5 above)
Comment (Christopher-Norman) [reply utils/data_version_control.py]: I think it is fine as it is for now

File: tests/test_utils/test_data_version_control.py [hunk 9]
```diff
@@ -23,3 +29,150 @@ def test_file_not_found_error_dvc(directory_for_testing):
...
+    directory_for_testing,
+):
+    """Test the _git_tag_exists function works as expected.
+
+    Args:
+        directory_for_testing (str): Path to testing directory.
+    """
+    data_path = os.path.join(directory_for_testing, "data")
+    os.mkdir(data_path)
+    with mock.patch("utils.data_version_control.git.Repo") as mock_git_repo:
```
Comment (JonoCX) [inline tests/test_utils/test_data_version_control.py]: Nit: could be a fixture which yields the `mock_git_repo`

File: tests/test_steps/test_generic_steps/test_load_data_step.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        directory_for_testing (str): Mock testing directory.
+        sample_scraping_data (Tuple[DataFrame, DataFrame]): Mock DataFrames for saving/loading.
+    """
+    data_location = os.path.join(directory_for_testing, "data")
+    os.mkdir(data_location)
+    mind, nhs = sample_scraping_data
+    mind.to_csv(f"{data_location}/mind_data_raw.csv", index=False)
+    nhs.to_csv(f"{data_location}/nhs_data_raw.csv", index=False)
+
+    with patch(
```
Comment (JonoCX) [inline tests/test_steps/test_generic_steps/test_load_data_step.py]: Nit: these patches could be a fixture which yields the context back to testing function

**Review by JonoCX — APPROVED** [review]
Nice work - a couple of nits 🌮

**Review by JonoCX — APPROVED** [review]
LGTM - nice 🌯

**Pull request merged**